### PR TITLE
data: Explicitly disavow writing data-* attributes

### PR DESCRIPTION
**Pull request by gibson042**
Also includes general cleanup.

Fixes gh-1023

### CODE REVIEW

File: entries/data.xml [hunk 1]
```diff
@@ -76,46 +76,42 @@ $( "span:last" ).text( $( "div" ).data( "test" ).last );
...
       <h4 id="data-html5">
         <a href="#data-html5">HTML5 data-* Attributes</a>
       </h4>
-      <p>As of jQuery 1.4.3 <a href="https://johnresig.com/blog/html-5-data-attributes/">HTML 5 data- attributes</a> will be automatically pulled in to jQuery's data object. The treatment of attributes with embedded dashes was changed in jQuery 1.6 to conform to the <a href="https://www.w3.org/TR/html5/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes">W3C HTML5 specification</a>.</p>
+      <p>Since jQuery 1.4.3, <a href="https://html.spec.whatwg.org/multipage/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes">data-* attributes</a> are used to initialize jQuery data. An element's data-* attributes are retrieved the first time the <code>data()</code> method is invoked upon it, and then are no longer accessed or mutated (all values are stored internally by jQuery).</p>
```
Comment (mgol) [inline entries/data.xml]: What do you think about wrapping all `data-*` occurrences in code blocks?

File: entries/data.xml [hunk 2]
```diff
@@ -76,46 +76,42 @@ $( "span:last" ).text( $( "div" ).data( "test" ).last );
...
+      <p>Calling <code>.data()</code> with no parameters returns a JavaScript object containing each stored value as a property. Using the object directly to get or set values is faster than making individual calls to <code>.data()</code>:</p>
       <pre><code>
-alert( $( "body" ).data( "foo" ) ); // undefined
-$( "body" ).data( "bar", "foobar" );
-alert( $( "body" ).data( "bar" ) ); // foobar
-      </code></pre>
-      <p><strong>jQuery 3</strong> changes the behavior of this method to align it to the <a href="https://www.w3.org/TR/html5/dom.html#dom-dataset">Dataset API specifications</a>. Specifically, jQuery 3 transforms every two-character sequence of "-" (U+002D) followed by a lowercase ASCII letter by the uppercase version of the letter as per definition of <a href="https://www.w3.org/TR/html5/dom.html#dom-dataset">the algorithm of the Dataset API</a>. Writing a statement like <code>$( "body" ).data( { "my-name": "aValue" } ).data();</code> will return <code>{ myName: "aValue" }</code>.</p>
+var mydata = $( "#mydiv" ).data();
+mydata.count = 43;
+mydata.status = "embiggened";
```
[the remark this answers is not in the document]
Comment (mgol) [reply entries/data.xml]: Hmm, I see it was already documented, just in a different place... I'd still warn against such usage, even if supported in a way (are we testing that?).

**Comment by mgol** [comment]
@coldice could you have a look? The clarification related to #1023 is the second commit of the PR: https://github.com/jquery/api.jquery.com/pull/1112/commits/34b6d5247aa6e9e96c852e4a27b8c67de51e8fab

**Comment by coldice** [comment]
@mgol Looks good 👍 The change in line 80 to "[..] allows us to **read** data [..]" also clarifies it very well. Thank you.

File: entries/data.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (gibson042) [reply entries/data.xml]: There don't seem to be any tests for _mutating_ the object returned by `.data()` in [test/unit/data.js](https://github.com/jquery/jquery/blob/354f6036f251a3ce9b24cd7b228b4c7a79001520/test/unit/data.js)... I'll temper this example.

File: entries/data.xml (hunk 1 above)
Comment (gibson042) [reply entries/data.xml]: There is precedent for both `<code>`-wrapped and unwrapped attribute names throughout the API documentation, including both patterns on the same page in e.g. [`.attr()`](https://api.jquery.com/attr/). But I also like it better with wrapping. Done.

**Pull request merged**